### PR TITLE
Enable setting an initial position and maximization launch for Terminal 

### LINKED ISSUE #1043: Be able to set an initial position for the terminal
# Environment

```none
Windows build number: Microsoft Windows [Version 10.0.18362.86]
Windows Terminal version (if applicable): commit bbbd3e0323a82548416f40ddba
```

# Steps to reproduce

1. Start Windows Terminal
2. Move the Windows Terminal window
3. Close Windows Terminal
4. Start Windows Terminal

# Expected behavior

Window position is restored to were it was in step 3 above.

# Actual behavior

Window is opened at a default location.

### PR DESCRIPTION
**Pull request by KaiyuWang16**
This PR is for code changes as well as a spec for https://github.com/microsoft/terminal/issues/1043

 
No

## PR Checklist
* [x] Closes #1043 
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

This PR includes the code changes that enable users to set an initial position (top left corner) and launch maximized. There are some corner cases:

1. Multiple monitors. The user should be able to set the initial position to any monitors attached. For the monitors on the left side of the major monitor, the initial position values are negative. 

2. If the initial position is larger than the screen resolution and the window is off-screen, the current solution is to check if the top left corner of the window intersect with any monitors. If it is not, we set the initial position to the top left corner of the nearest monitor.

3.  If the user wants to launch maximized and provides an initial position, we launch the maximized window on the monitor where the position is located. 

Testing:

To test:
1. Check-out this branch and build on VS2019
2. Launch Terminal, and open Settings. Then close the terminal. 
3. Add the following setting into Json settings file as part of "globals", just after "initialRows":
                   "initialPosition": "1000, 1000",
                   "launchMode": "default"

My test data: 
I have already tested with the following variables:
                                1. showTabsInTitlebar true or false
                                2. The initial position of the top left corner of the window
                                3. Whether to launch maximized
                                4. The DPI of the monitor

Test data combination:

Non-client island window (showTabsInTitlebar true)

1. Three monitors with the same DPI (100%), left, middle and right, with the middle one as the primary, resolution: 1980 * 1200, 1920 * 1200, 1920 * 1080
    launchMode: default
                          In-Screen test: (0, 0), (1000, 500), (2000, 300), (-1000, 400), (-100, 200), (-2000, 100),
                                                  (0, 1119)
                          out-of-screen:
                                                  (200, -200): initialize to (0, 0)
                                                  (200, 1500): initialize to (0, 0)
                                                  (2000, -200): initialize to (1920, 0)
                                                  (2500, 2000): initialize to (1920, 0)
                                                  (4000 100): initialize to (1920, 0)
                                                  (-1000, -100): initialize to (-1920, 0)
                                                  (-3000, 100): initialize to (-1920, 0)
                                                  (10000, -10000): initialize to (1920, 0)
                                                  (-10000, 10000): initialize to (-1920, 0)
                                                  (0, -10000): initialize to (0, 0)
                                                  (0, -1):  initialize to (0, 0)
                                                  (0, 1200):  initialize to (0, 0)
    launch mode: maximize
                                   (100, 100)
                                   (-1000, 100): On the left monitor
                                   (0, -2000): On the primary monitor
                                   (10000, 10000): On the primary monitor

2. Left monitor 200% DPI, primary monitor 100% DPI
    In screen: (-1900, 100), (-3000, 100), (-1000, 100)
    our-of-screen: (-8000, 100): initialize at (-1920, 0)
    launch Maximized:  (-100, 100): launch maximized on the left monitor correctly

3. Left monitor 100% DPI, primary monitor 200% DPI
    In-screen: (-1900, 100), (300, 100), (-800, 100), (-200, 100)
    out-of-screen: (-3000, 100): initialize at (-1920, 0)
    launch maximized: (100, 100), (-1000, 100)

For client island window, the test data is the same as above. 

Issues:

1. If we set the initial position on the monitor with a different DPI as the primary monitor, and the window "lays" across two monitors, then the window still renders as it is on the primary monitor. The size of the window is correct.

### CODE REVIEW

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 1]
```diff
@@ -208,6 +257,8 @@ void AppHost::_HandleCreateWindow(const HWND hwnd, const RECT proposedRect)
         TraceLoggingDescription("Event emitted upon creating the application window"),
         TraceLoggingKeyword(MICROSOFT_KEYWORD_MEASURES),
         TelemetryPrivacyDataTag(PDT_ProductAndServicePerformance));
+
+    return _app.GetLaunchMode();
```
Comment (codendone) [inline src/cascadia/WindowsTerminal/AppHost.cpp]: This can just be "return launchMode;"

File: doc/specs/#1043 - Set the initial position of the Terminal/spec.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
+---
```
Comment (codendone) [inline doc/specs/#1043 - Set the initial position of the Terminal/spec.md]: This file should be removed from this change, since it is part of the separate spec PR, right?

**Comment by miniksa** [comment]
> 2. For client island window,  if we set the initial position to the top left corner of a monitor,  the left edge of the window does not coincide with the left edge of the screen. We may need to adjust the size of the window.

Is it by chance off by 8 pixels from the left on a 100% DPI monitor? If so, I'm pretty sure there's a method or window metric hanging around somewhere in the system that will give you that figure of the "invisible window border" so you can offset your position. I know that we talked through this once in the past because conhost.exe set to 0,0 is offset slightly from the left. I don't think we mitigated it, but it is definitely possible to lookup the window metric. If this is the situation, let me know and I'll try to find my notes on the right way to do it.

File: doc/specs/#1043 - Set the initial position of the Terminal/spec.md (hunk 2 above)
Comment (miniksa) [reply doc/specs/#1043 - Set the initial position of the Terminal/spec.md]: Yes, please do not include the spec in both PRs.

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp [hunk 3]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
+#include <WinUser.h>
```
Comment (miniksa) [inline src/cascadia/TerminalApp/GlobalAppSettings.cpp]: This should probably go into the pch.h if it isn't there already.

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp [hunk 4]
```diff
@@ -37,11 +40,16 @@ GlobalAppSettings::GlobalAppSettings() :
     _alwaysShowTabs{ true },
     _initialRows{ DEFAULT_ROWS },
     _initialCols{ DEFAULT_COLS },
+    _initialX{ 0 },
+    _initialY{ 0 },
+    _useDefaultInitialX{ true },
```
Comment (miniksa) [inline src/cascadia/TerminalApp/GlobalAppSettings.cpp]: Shouldn't these be false by default unless set by the load of preferences?

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 5]
```diff
@@ -126,70 +126,115 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
...
+        GetMonitorInfoA(hmon, &monitorInfo);
+
+        // This API guarantees that dpix and dpiy will be equal, but neither is an
+        // optional parameter so give two UINTs.
+        UINT dpix = USER_DEFAULT_SCREEN_DPI;
+        UINT dpiy = USER_DEFAULT_SCREEN_DPI;
+        // If this fails, we'll use the default of 96.
+        GetDpiForMonitor(hmon, MDT_EFFECTIVE_DPI, &dpix, &dpiy);
+
+        // We need to check if the top left point of the titlebar of the window is winthin any screen
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/AppHost.cpp]:
```suggestion
        // We need to check if the top left point of the titlebar of the window is within any screen
```

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 6]
```diff
@@ -126,70 +126,115 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
...
+                             .ToRect();
+
+        // Get the size of a window we'd need to host that client rect. This will
+        // add the titlebar space.
+        if (_useNonClientArea)
+        {
+            // If we're in NC tabs mode, do the math ourselves. Get the margins
+            // we're using for the window - this will include the size of the
+            // titlebar content.
+            auto pNcWindow = static_cast<NonClientIslandWindow*>(_window.get());
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/AppHost.cpp]: I think this can be const.

```suggestion
            const auto pNcWindow = static_cast<NonClientIslandWindow*>(_window.get());
```

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 7]
```diff
@@ -126,70 +126,115 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
...
+                // size of our window, which will be at least close.
+                LOG_LAST_ERROR();
+                nonClient = Viewport::FromDimensions({ _currentWidth,
+                                                       _currentHeight })
+                                .ToRect();
+            }
+        }
+
+        // Calculate the maximum size the window could have without hangs-off
+        long horizontalMaxLength = std::abs(proposedRect.left - monitorInfo.rcWork.right);
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/AppHost.cpp]: These look const-able too.

```suggestion
        const long horizontalMaxLength = std::abs(proposedRect.left - monitorInfo.rcWork.right);
```

File: src/cascadia/WindowsTerminal/BaseWindow.h [hunk 8]
```diff
@@ -215,6 +215,15 @@ class BaseWindow
         PostMessageW(_window.get(), CM_UPDATE_TITLE, 0, reinterpret_cast<LPARAM>(nullptr));
     }
 
+    // Method Description:
+    // Reset the current dpi of the window. This method is only called after we change the
+    // initial launch position. This makes sure the dpi is consistent with the monitor on which
+    // the window will launch
+    void RefreshCurrentDPI()
+    {
+        _currentDpi = GetDpiForWindow(this->_window.get());
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/BaseWindow.h]: removed

File: src/cascadia/TerminalApp/GlobalAppSettings.h [hunk 9]
```diff
@@ -75,6 +86,11 @@ class TerminalApp::GlobalAppSettings final
     int32_t _initialRows;
     int32_t _initialCols;
 
+    int _initialX;
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.h]: changed, mistake

File: src/cascadia/TerminalApp/GlobalAppSettings.h [hunk 10]
```diff
@@ -75,6 +86,11 @@ class TerminalApp::GlobalAppSettings final
     int32_t _initialRows;
     int32_t _initialCols;
 
+    int _initialX;
+    int32_t _initialY;
+    bool _useDefaultInitialX;
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.h]: changed

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp [hunk 11]
```diff
@@ -134,6 +151,32 @@ void GlobalAppSettings::SetShowTabsInTitlebar(const bool showTabsInTitlebar) noe
 {
     _showTabsInTitlebar = showTabsInTitlebar;
 }
+
+int32_t GlobalAppSettings::GetInitialX() const noexcept
+{
+    return _initialX;
+}
+void GlobalAppSettings::SetInitialX(const int32_t initialX) noexcept
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp]: changed

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp [hunk 12]
```diff
@@ -164,11 +208,14 @@ Json::Value GlobalAppSettings::ToJson() const
     jsonObject[JsonKey(DefaultProfileKey)] = winrt::to_string(Utils::GuidToString(_defaultProfile));
     jsonObject[JsonKey(InitialRowsKey)] = _initialRows;
     jsonObject[JsonKey(InitialColsKey)] = _initialCols;
+    jsonObject[JsonKey(InitialXKey)] = _initialX;
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp]: Add if check using _isInitialX/YSet

File: src/cascadia/TerminalApp/App.cpp [hunk 13]
```diff
@@ -342,6 +342,58 @@ namespace winrt::TerminalApp::implementation
...
+        {
+            // Load settings if we haven't already
+            LoadSettings();
+        }
+
+        winrt::Windows::Foundation::Point point((float)defaultInitialX, (float)defaultInitialY);
+
+        if (!_settings->GlobalSettings().GetUseDefaultInitialX())
+        {
+            point.X = (float)_settings->GlobalSettings().GetInitialX();
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/App.cpp]: changed

File: src/cascadia/TerminalApp/App.h [hunk 14]
```diff
@@ -31,6 +31,8 @@ namespace winrt::TerminalApp::implementation
         void LoadSettings();
 
         Windows::Foundation::Point GetLaunchDimensions(uint32_t dpi);
+        winrt::Windows::Foundation::Point GetLaunchInitialPositions(int defaultInitialX, int defaultInitialY);
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/App.h]: changed

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 15]
```diff
@@ -126,70 +126,115 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
...
-        nonClient.top = 0;
-        nonClient.right = margins.cxLeftWidth + nonClient.right + margins.cxRightWidth;
-        nonClient.bottom = margins.cyTopHeight + nonClient.bottom + margins.cyBottomHeight;
-    }
-    else
+    winrt::hstring launchMode = _app.GetLaunchMode();
+
+    // Acquire the actual intial position
+    winrt::Windows::Foundation::Point initialPosition = _app.GetLaunchInitialPositions(proposedRect.left, proposedRect.top);
+    proposedRect.left = (long)initialPosition.X;
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: use narrow_cast now

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 16]
```diff
@@ -126,70 +126,115 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
...
     {
-        bool succeeded = AdjustWindowRectExForDpi(&nonClient, WS_OVERLAPPEDWINDOW, false, 0, dpix);
-        if (!succeeded)
+        // Find nearest montitor.
+        HMONITOR hmon = MonitorFromRect(&proposedRect, MONITOR_DEFAULTTONEAREST);
+
+        // Get nearest monitor information
+        MONITORINFO monitorInfo;
+        monitorInfo.cbSize = sizeof(MONITORINFO);
+        GetMonitorInfoA(hmon, &monitorInfo);
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: removed

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 17]
```diff
@@ -126,70 +126,115 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
...
+        offScreenTestRect.right = offScreenTestRect.left + 1;
+        offScreenTestRect.bottom = offScreenTestRect.top + 1;
+
+        bool isTitlebarIntersectWithMonitors = false;
+        EnumDisplayMonitors(
+            nullptr, &offScreenTestRect, [](HMONITOR, HDC, LPRECT, LPARAM lParam) -> BOOL {
+                auto intersectWithMonitor = reinterpret_cast<bool*>(lParam);
+                *intersectWithMonitor = true;
+                // Continue the enumeration
+                return TRUE;
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: Oh, we do not need to continue, changed

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 18]
```diff
@@ -126,70 +126,115 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
...
+                // the error and go on. We'll use whatever the control proposed as the
+                // size of our window, which will be at least close.
+                LOG_LAST_ERROR();
+                nonClient = Viewport::FromDimensions({ _currentWidth,
+                                                       _currentHeight })
+                                .ToRect();
+            }
+        }
+
+        // Calculate the maximum size the window could have without hangs-off
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: Rich Turner said that If, after figuring out the starting position, the Terminal hangs off the screen, the Terminal’s dimensions should be reduced to ensure that the Terminal fits on-screen. But yes, this is an existing issue. Shall I ask Rich to verify if we want this to be included this time ?

File: src/cascadia/WindowsTerminal/AppHost.cpp (hunk 1 above)
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: changed

File: src/cascadia/WindowsTerminal/AppHost.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (codendone) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: This behavior is not consistent with what I see today in a PowerShell:  If I set a wide default width (which seems to get limited to the width of the screen), I can still set an offset which puts some of it offscreen.  In a multi-mon scenario, I think it is even less desirable to force the window entirely on one monitor.  I suggest not doing this change, at least not unless/until all scenarios have been considered.

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp (hunk 3 above)
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp]: Seems that we do not need this header. removed.

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp (hunk 4 above)
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp]: I have changed the flag to isInitialX/YSet, and the default value is false. For the _initialX/Y, we actually won't use this default value, I wonder if you think we can just initialize them with {} ?

File: src/cascadia/WindowsTerminal/AppHost.cpp (hunk 6 above)
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: changed

File: src/cascadia/WindowsTerminal/AppHost.cpp (hunk 7 above)
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: changed

File: src/cascadia/TerminalApp/App.idl [hunk 19]
```diff
@@ -24,6 +24,8 @@ namespace TerminalApp
         String Title { get; };
 
         Windows.Foundation.Point GetLaunchDimensions(UInt32 dpi);
+        Windows.Foundation.Point GetLaunchInitialPositions(UInt64 defaultInitialX, UInt64 defaultInitialY);
+        String GetLaunchMode();
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/App.idl]: I would put the definition of the Enum in App because this enum needs to be in the idl file.

**Comment by KaiyuWang16** [comment]
> For client island window,  if we set the initial position to the top left corner of a monitor,  the left edge of the window does not coincide with the left edge of the screen. We may need to adjust the size of the window.
> 
> 
> Is it by chance off by 8 pixels from the left on a 100% DPI monitor? If so, I'm pretty sure there's a method or window metric hanging around somewhere in the system that will give you that figure of the "invisible window border" so you can offset your position. I know that we talked through this once in the past because conhost.exe set to 0,0 is offset slightly from the left. I don't think we mitigated it, but it is definitely possible to lookup the window metric. If this is the situation, let me know and I'll try to find my notes on the right way to do it.

Yes, if I move the window left for 8 pixies, if will match the left edge of the screen. Griese also told me that there is a 8 pixies border on win32 windows, but win10 windows only have 1 pixel, and this introduces an invisible border around the Terminal window.

File: src/cascadia/TerminalApp/App.cpp [hunk 20]
```diff
@@ -342,6 +342,65 @@ namespace winrt::TerminalApp::implementation
...
+    {
+        if (!_loadedInitialSettings)
+        {
+            // Load settings if we haven't already
+            LoadSettings();
+        }
+
+        hstring launchModeText = _settings->GlobalSettings().GetLaunchMode();
+        LaunchMode launchMode = LaunchMode::DefaultMode;
+        if (launchModeText == L"maximized")
```
Comment (codendone) [inline src/cascadia/TerminalApp/App.cpp]: I'd like to see the conversion to/from string happening in GlobalAppSettings.cpp, similar to what it does for _requestedTheme.

Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/App.cpp]: changed

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp [hunk 21]
```diff
@@ -280,3 +354,38 @@ std::wstring_view GlobalAppSettings::_SerializeTheme(const ElementTheme theme) n
...
+    return LaunchMode::DefaultMode;
+}
+
+// Method Description:
+// - Helper function for converting a LaunchMode to its corresponding string
+//   value.
+// Arguments:
+// - launchMode: The enum value to convert to a string.
+// Return Value:
+// - The string value for the given CursorStyle
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp]: changed

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 22]
```diff
@@ -126,70 +126,118 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
...
+                // If we failed to get the correct window size for whatever reason, log
+                // the error and go on. We'll use whatever the control proposed as the
+                // size of our window, which will be at least close.
+                LOG_LAST_ERROR();
+                nonClient = Viewport::FromDimensions({ _currentWidth,
+                                                       _currentHeight })
+                                .ToRect();
+            }
+
+            // For client island scenario, there is an invisible border of 8 pixals.
```
[the remark this answers is not in the document]
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: changed

File: doc/specs/#1043 - Set the initial position of the Terminal/spec.md (hunk 2 above)
Comment (KaiyuWang16) [reply doc/specs/#1043 - Set the initial position of the Terminal/spec.md]: removed

File: src/cascadia/TerminalApp/GlobalAppSettings.h [hunk 23]
```diff
@@ -58,6 +58,17 @@ class TerminalApp::GlobalAppSettings final
     bool GetCopyOnSelect() const noexcept;
     void SetCopyOnSelect(const bool copyOnSelect) noexcept;
 
+    int32_t GetInitialX() const noexcept;
+    void SetInitialX(const int32_t initialX) noexcept;
+    bool GetIsInitialXSet() const noexcept;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/GlobalAppSettings.h]: Can we change `GetIsInitial*Set()` to `IsInitial*Set()`? It'll read better when it gets used.

Comment (carlos-zamora) [reply src/cascadia/TerminalApp/GlobalAppSettings.h]: Also change `_initialX` and `_initialY` from `int32_t` to `std::optional<int32_t>`. Maybe then we wouldn't even need these two functions.

Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.h]: Changed

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp (hunk 4 above)
Comment (miniksa) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp]: I just meant `false` instead of `true` which it looks like you did in the latest revision.

File: src/cascadia/WindowsTerminal/IslandWindow.h [hunk 24]
```diff
@@ -8,6 +8,7 @@
 #include <winrt/Microsoft.Terminal.TerminalControl.h>
 #include <winrt/TerminalApp.h>
 #include "../../cascadia/inc/cppwinrt_utils.h"
+#include <windowsx.h>
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/IslandWindow.h]: This probably belongs in precomp.h

Comment (miniksa) [inline src/cascadia/WindowsTerminal/IslandWindow.h]: What specifically is using windowsx.h in this header? if it's only something in the cpp file, it should probably be in precomp or the cpp file.

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 25]
```diff
@@ -126,70 +126,118 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
 //   create. We'll use this rect to determine which monitor the window is about
 //   to appear on.
 // Return Value:
-// - <none>
-void AppHost::_HandleCreateWindow(const HWND hwnd, const RECT proposedRect)
+// - A LaunchMode enum that indicates the launch mode
+winrt::TerminalApp::LaunchMode AppHost::_HandleCreateWindow(const HWND hwnd, RECT proposedRect)
```
Comment (miniksa) [inline src/cascadia/WindowsTerminal/AppHost.cpp]: I don't know how to feel about returning the launch mode here. Was this a @DHowett-MSFT or @zadjii-msft request? It seems a little odd to me.

File: src/cascadia/WindowsTerminal/IslandWindow.h (hunk 24 above)
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/IslandWindow.h]: moved to pch.h

File: src/cascadia/WindowsTerminal/AppHost.cpp (hunk 25 above)
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: Changed. Now pass LaunchMode as reference to _HandleCreateWindow.

File: src/cascadia/TerminalApp/GlobalAppSettings.h [hunk 26]
```diff
@@ -84,8 +100,24 @@ class TerminalApp::GlobalAppSettings final
...
+    winrt::TerminalApp::LaunchMode _launchMode;
+
     static winrt::Windows::UI::Xaml::ElementTheme _ParseTheme(const std::wstring& themeString) noexcept;
     static std::wstring_view _SerializeTheme(const winrt::Windows::UI::Xaml::ElementTheme theme) noexcept;
 
+    static void _ParseInitialPosition(const std::wstring& initialPosition,
+                                      int32_t& initialX,
+                                      bool& isInitialXSet,
+                                      int32_t& initialY,
+                                      bool& isInitialYSet) noexcept;
```
Comment (codendone) [inline src/cascadia/TerminalApp/GlobalAppSettings.h]: _isInitialXSet and _isInitialYSet don't need to be tracked separately now that _initialX/Y are std::optional<>.  Can you eliminate those bools, pass the parameters as std::optional<> to these functions, and use _initialX.has_value() in any place which wants to know if it is set?

Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.h]: Removed the useless bools.

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp [hunk 27]
```diff
@@ -134,6 +151,27 @@ void GlobalAppSettings::SetShowTabsInTitlebar(const bool showTabsInTitlebar) noe
...
 }
+
+std::optional<int32_t> GlobalAppSettings::GetInitialX() const noexcept
+{
+    return _initialX;
+}
+
+void GlobalAppSettings::SetInitialX(const int32_t initialX) noexcept
+{
+    _initialX.value() = initialX;
```
Comment (codendone) [inline src/cascadia/TerminalApp/GlobalAppSettings.cpp]: Why did you change this from "_initialX = initialX"?  Does the new code really work?  It looks like std::optional will throw an exception if .value() is called when it doesn't yet have a value.  Should these use emplace(), like your Parse function does?  It looks like there aren't any callers of SetInitialX/Y -- should these functions be removed if they aren't currently needed/

Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp]: Sorry for this mistake. I should use emplace here. And I removed them because they are not used anywhere.

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp [hunk 28]
```diff
@@ -280,3 +319,115 @@ std::wstring_view GlobalAppSettings::_SerializeTheme(const ElementTheme theme) n
...
+    size_t* idx = nullptr;
+
+    // Get initial position values till we run out of delimiter separated values in the stream
+    // or we hit max number of allowable values (= 2)
+    // Non-numeral values or empty string will be caught as exception and we do not assign them
+    for (; std::getline(tokenStream, token, singleCharDelim) && (initialPosIndex < 2); initialPosIndex++)
+    {
+        try
+        {
+            int32_t position = std::stoi(token, idx);
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/GlobalAppSettings.cpp]: since you are never using `idx` later, you can use the single-argument version of `std::stoi`

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp [hunk 29]
```diff
@@ -280,3 +319,115 @@ std::wstring_view GlobalAppSettings::_SerializeTheme(const ElementTheme theme) n
...
+//   value.
+// Arguments:
+// - initialX: reference to the _initialX member
+//   initialY: reference to the _initialY member
+// Return Value:
+// - The concatenated string for the the current initialX and initialY
+std::string GlobalAppSettings::_SerializeInitialPosition(const std::optional<int32_t>& initialX,
+                                                         const std::optional<int32_t>& initialY) noexcept
+{
+    std::string serializedInitialPos = "(";
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/GlobalAppSettings.cpp]: the deserializer doesn't use `(` or `)`, should we just use `100, 200` instead? I don't think the user should have to type `(100, 200)` :smile:

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp [hunk 30]
```diff
@@ -280,3 +319,115 @@ std::wstring_view GlobalAppSettings::_SerializeTheme(const ElementTheme theme) n
...
+//   initialX: reference to the _initialX member
+//   initialY: reference to the _initialY member
+// Return Value:
+// - None
+void GlobalAppSettings::_ParseInitialPosition(const std::wstring& initialPosition,
+                                              std::optional<int32_t>& initialX,
+                                              std::optional<int32_t>& initialY) noexcept
+{
+    const wchar_t singleCharDelim = L',';
+    std::wstringstream tokenStream(initialPosition.c_str());
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/GlobalAppSettings.cpp]: you do not need to use `c_str()` to construct a wstringstream from a wstring, you can just use the wstring directly.

```
std::wstringstream tokenStream{initialPosition};
```

File: src/cascadia/WindowsTerminal/AppHost.cpp [hunk 31]
```diff
@@ -125,71 +126,120 @@ void AppHost::LastTabClosed(const winrt::Windows::Foundation::IInspectable& /*se
...
+                                                       _currentHeight })
+                                .ToRect();
+            }
+
+            // For client island scenario, there is an invisible border of 8 pixels.
+            // We need to remove this border to guarantee the left edge of the window
+            // coincides with the screen
+            const auto pCWindow = static_cast<IslandWindow*>(_window.get());
+            const RECT frame = pCWindow->GetFrameBorderMargins(dpix);
+            proposedRect.left -= (frame.left * -1);
```
Comment (DHowett-MSFT) [inline src/cascadia/WindowsTerminal/AppHost.cpp]: "left minus left*-1" sounds like "left + left"; can this be simplified?

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp (hunk 28 above)
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp]: removed

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp (hunk 29 above)
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp]: removed, thanks for this reminding !

File: src/cascadia/TerminalApp/GlobalAppSettings.cpp (hunk 30 above)
Comment (KaiyuWang16) [reply src/cascadia/TerminalApp/GlobalAppSettings.cpp]: removed c_str()

File: src/cascadia/WindowsTerminal/AppHost.cpp (hunk 31 above)
Comment (KaiyuWang16) [reply src/cascadia/WindowsTerminal/AppHost.cpp]: Sorry for this stupid mistake, changed

**Comment by DHowett-MSFT** [comment]
I will merge with the following message when the builds complete.

```
This PR includes the code changes that enable users to set an initial position
(top left corner) and launch maximized. There are some corner cases:

1. Multiple monitors. The user should be able to set the initial position to
any monitors attached. For the monitors on the left side of the major monitor,
the initial position values are negative.

2. If the initial position is larger than the screen resolution and the window
is off-screen, the current solution is to check if the top left corner of the
window intersect with any monitors. If it is not, we set the initial position
to the top left corner of the nearest monitor.

3. If the user wants to launch maximized and provides an initial position, we
launch the maximized window on the monitor where the position is located.

# Testing

To test:
1. Check-out this branch and build on VS2019
2. Launch Terminal, and open Settings. Then close the terminal.
3. Add the following setting into Json settings file as part of "globals", just
after "initialRows":
  "initialPosition": "1000, 1000",
  "launchMode": "default"

My test data:
I have already tested with the following variables:
  1. showTabsInTitlebar true or false
  2. The initial position of the top left corner of the window
  3. Whether to launch maximized
  4. The DPI of the monitor

Test data combination:

Non-client island window (showTabsInTitlebar true)

1. Three monitors with the same DPI (100%), left, middle and right, with the
middle one as the primary, resolution: 1980 * 1200, 1920 * 1200, 1920 * 1080
    launchMode: default
      In-Screen test: (0, 0), (1000, 500), (2000, 300), (-1000, 400),
        (-100, 200), (-2000, 100), (0, 1119)
      out-of-screen:
        (200, -200): initialize to (0, 0)
        (200, 1500): initialize to (0, 0)
        (2000, -200): initialize to (1920, 0)
        (2500, 2000): initialize to (1920, 0)
        (4000 100): initialize to (1920, 0)
        (-1000, -100): initialize to (-1920, 0)
        (-3000, 100): initialize to (-1920, 0)
        (10000, -10000): initialize to (1920, 0)
        (-10000, 10000): initialize to (-1920, 0)
        (0, -10000): initialize to (0, 0)
        (0, -1):  initialize to (0, 0)
        (0, 1200):  initialize to (0, 0)
    launch mode: maximize
        (100, 100)
        (-1000, 100): On the left monitor
        (0, -2000): On the primary monitor
        (10000, 10000): On the primary monitor


2. Left monitor 200% DPI, primary monitor 100% DPI
    In screen: (-1900, 100), (-3000, 100), (-1000, 100)
    our-of-screen: (-8000, 100): initialize at (-1920, 0)
    launch Maximized:  (-100, 100): launch maximized on the left monitor
      correctly

3. Left monitor 100% DPI, primary monitor 200% DPI
    In-screen: (-1900, 100), (300, 100), (-800, 100), (-200, 100)
    out-of-screen: (-3000, 100): initialize at (-1920, 0)
    launch maximized: (100, 100), (-1000, 100)

For client island window, the test data is the same as above.

Issues:

1. If we set the initial position on the monitor with a different DPI as the
primary monitor, and the window "lays" across two monitors, then the window
still renders as it is on the primary monitor. The size of the window is
correct.

Closes #1043
```
